posts/Jan/2022-04-29-mapping-boolean-types-in-hibernate-orm-6.adoc

```diff
@@ -0,0 +1,149 @@
+:awestruct-layout: blog-post
+:icons: font
+Jan Schatteman <jschatte@redhat.com>
+
+By default, Hibernate maps a Boolean value to a BOOLEAN column. However, not all databases have a built-in BOOLEAN datatype and not all applications will want to use BOOLEAN (legacy schema e.g.).
+
+Hibernate provides multiple ways to map Boolean value to other SQL types; this post explores the most common ways.
+
+.Implicit Boolean mapping
+====
+[source, JAVA, indent=0]
+----
+
+    @Basic
+    private boolean aBoolean;
+
+----
+====
+How Hibernate chooses to map a boolean implicitly is determined by a series of cascading checks:
+
+1. It first checks whether the link:{javadoc-AvailableSettings}#PREFERRED_BOOLEAN_JDBC_TYPE[hibernate.type.preferred_boolean_jdbc_type] setting has been specified and, if so, uses it;
+2. Next it consults the link:{javadoc-Dialect}#getPreferredSqlTypeCodeForBoolean[Dialect] to see if it reports a preferred SQL type and, if so, uses it;
+3. Otherwise it will use link:{javadoc-Types}#BOOLEAN[Types.BOOLEAN]
+
+This default, 'implicit' mapping can be modified in several ways.
+
+== JPA compliant, using AttributeConverter
+
+Out of the box, Hibernate provides 3 `AttributeConverters`
+
+* YesNoConverter encodes a boolean value as 'Y' or 'N',
+* TrueFalseConverter encodes a boolean value as 'T' or 'F', and
+* NumericBooleanConverter encodes the value as an integer, 1 for true, and 0 for false.
+
+.Using built-in AttributeConverters
+====
+[source, JAVA, indent=0]
+----
+// this will get mapped to CHAR or NCHAR
+@Basic
+@Convert(converter = org.hibernate.type.YesNoConverter.class)
+boolean convertedYesNo;
+
+// this will get mapped to CHAR or NCHAR
+@Basic
+@Convert(converter = org.hibernate.type.TrueFalseConverter.class)
+boolean convertedTrueFalse;
+
+// this will get mapped to TINYINT
+@Basic
+@Convert(converter = org.hibernate.type.NumericBooleanConverter.class)
+boolean convertedNumeric;
+----
+====
+In the above examples, the converter will only be applied to the specific attribute that is annotated. So, what if you want to apply such a converter to all Boolean attributes in your application? Until now this would have required you to either define a custom `AttributeConverter` (see below), by, for example, extending either one of the available ones described above, and setting `autoApply` on it to `true`.
+
+NOTE: in impending ORM 6.1 release, a new `@ConverterRegistration` was added to make this much easier. It allows modifying certain aspects of already defined `AttributeConverters`, such as the application domain (`autoApply`) or change the data type the conversion should be applied on to another compatible data type (e.g. a subclass of the original data type).
+
+.Customizing built-in AttributeConverters
+====
+[source, JAVA, indent=0]
+----
+// this will apply the converter to all Boolean attributes
+@Basic
+@ConverterRegistration( converter = org.hibernate.type.YesNoConverter.class, autoApply = true )
```

Comment:
This is not allow on fields or methods.  Only packages and types (classes, etc).

Reply:
UGH, how silly, copy-paste overlook. Will correct